Improve tooltip performance

connection-reset opened the pull request:
Tooltips aren't that performant and it's even worse when there are many.
I replaced the linear search in findHoverIndexFromData with binary search, which is faster in my benchmarks:
```
1000 data points
random posX, not necessarily in array
new, 1000000 calls, 125ms
old, 1000000 calls, 1772ms
posX in array
new, 1000000 calls, 103ms
old, 1000000 calls, 1405ms
```

Tooltips are only shown if they are in the browsers viewport.
![tooltip](https://cloud.githubusercontent.com/assets/6831775/22102325/604d59d2-de37-11e6-83f3-df0cab8c702d.gif)

I can't tell if they are actually faster now or if it's just placebo.

Commit:
Only render tooltip if position is in viewport

Commit:
Speed up tooltip value retrieval

apirek committed:
Clear highlight when clearing tooltip and crosshair

torkelo commented:
This looks great! 

strange about you noticing it's slow however, what browser are you using? Even with using shared tooltip on multiple graphs (with multiple high resolution series) its butter smooth in chrome for me. 
https://twitter.com/torkelo/status/806095071753990145

connection-reset commented:
Firefox 50.1 on Linux, it lags about half a second behind (a bit less now but still noticable).
Of course firefox just isn't as fast as chrome but I'd still like grafana to feel as good in firefox as it does in chrome.

apirek committed:
Fix tooltips sometimes rendering outside of graph panel

I noticed tooltips on other panels rendering outside of their panel or
even the browser window when hovering on the very edge of graphs with
relative time overrides.

torkelo commented:
Yea, have noticed that Grafana performance in Firefox is generally a lot slower than chrome, in some cases by vast margin. Even IE Egde is a lot faster

connection-reset commented:
I'd like to look for more performance / general improvements but I've got exams coming up so I can't work on this for a while.

torkelo commented:
merged this, but could not notice ANY performance difference in Firefox (on linux) with 10 graphs with 15 series each, but maybe there is noticeable difference on laptops.
